src/EFCore.PG/Query/Internal/NpgsqlQuerySqlGenerator.cs

```diff
@@ -464,62 +463,55 @@ public virtual Expression VisitCustomUnary([NotNull] SqlCustomUnaryExpression ex
+        /// <param name="e">The expression.</param>
         /// <returns>
         /// An <see cref="Expression"/>.
         /// </returns>
         [NotNull]
-        public virtual Expression VisitPgFunction([NotNull] PgFunctionExpression sqlFunctionExpression)
+        public virtual Expression VisitPgFunction([NotNull] PgFunctionExpression e)
         {
-            if (!string.IsNullOrEmpty(sqlFunctionExpression.Schema))
+            Check.NotNull(e, nameof(e));
+
+            if (e.IsBuiltIn)
             {
-                Sql
-                    .Append(_sqlGenerationHelper.DelimitIdentifier(sqlFunctionExpression.Schema))
-                    .Append(".");
+                Sql.Append(e.Name);
             }
-
-            // TODO: Quote user-defined function names with upper-case
-            Sql
-                .Append(sqlFunctionExpression.FunctionName)
-                .Append("(");
-
-            GenerateList(sqlFunctionExpression.PositionalArguments, e => Visit(e));
-
-            var hasArguments = sqlFunctionExpression.PositionalArguments.Count > 0 && sqlFunctionExpression.NamedArguments.Count > 0;
-
-            foreach (var kv in sqlFunctionExpression.NamedArguments)
+            else
             {
-                if (hasArguments)
-                    Sql.Append(", ");
-                else
-                    hasArguments = true;
+                if (!string.IsNullOrEmpty(e.Schema))
+                {
+                    Sql
+                        .Append(_sqlGenerationHelper.DelimitIdentifier(e.Schema))
+                        .Append(".");
+                }
 
-                Sql.Append(kv.Key).Append(" => ");
+                // TODO: Quote user-defined function names with upper-case (also for regular SqlFunctionExpression)
 
-                Visit(kv.Value);
+                Sql.Append(_sqlGenerationHelper.DelimitIdentifier(e.Name));
             }
 
-            Sql.Append(")");
-
-            return sqlFunctionExpression;
-        }
-
-        void GenerateList<T>(
-            IReadOnlyList<T> items,
-            Action<T> generationAction,
-            Action<IRelationalCommandBuilder> joinAction = null)
-        {
-            joinAction ??= (isb => isb.Append(", "));
+            Sql.Append("(");
 
-            for (var i = 0; i < items.Count; i++)
+            for (var i = 0; i < e.Arguments.Count; i++)
             {
-                if (i > 0)
+                if (i < e.ArgumentNames.Count && e.ArgumentNames[i] != null)
                 {
-                    joinAction(Sql);
+                    Sql
+                        .Append(e.ArgumentNames[i])
+                        .Append(" => ");
                 }
 
-                generationAction(items[i]);
+                Visit(e.Arguments[i]);
+
+                if (i < e.Arguments.Count - 1)
+                    Sql.Append(i < e.ArgumentSeparators.Count && e.ArgumentSeparators[i] != null
+                        ? $" {e.ArgumentSeparators[i]} "
```

Comment:
If perf matters I would recommend to use sting concatenation here.

Reply:
Perf doesn't really matter here (this is compilation-only, not runtime), but I was under the impression this was the same?

Reply:
Sorry, in case when an argument is a string, the compiler generates different code. It uses concatenation, but when there is a non-string parameter it will use formatting.

Reply:
Thanks for clarifying, good to know.